### PR TITLE
Infrastructure for ReactiveMessageHandler

### PR DESCRIPTION
**Pull request by artembilan**
We have now a `ReactiveMongoDbStoringMessageHandler` which implements
a `ReactiveMessageHandler`, but not a `MessageHandler` for possible
deferred subscriptions to the returned Reactor type

We don't have a proper application context processing for this
new type of message handlers

* Change a  `ConsumerEndpointFactoryBean` to apply an `MH` and `RMH`
as possible types for handler
* Introduce a `ReactiveMessageHandlerAdapter` to wrap an `RMH`
into a `MH` for synchronous calls in the regular consumer endpoints
* Wrap an `RMH` into a `ReactiveMessageHandlerAdapter` for regular
endpoints and unwrap for `ReactiveStreamsConsumer`
* Add `RMH`-based ctor into `ReactiveStreamsConsumer` for target
reactive streams composition (`flatMap()` on the `RMH`)
* Remove a `DelegatingSubscriber` from the `ReactiveStreamsConsumer`
in favor of direct calls from the `doOnSubscribe()`, `doOnComplete()`
& `doOnNext()`
* Add an `onErrorContinue()` to handle per-message errors, but don't
cancel the whole source `Publisher`
* Use `Disposable` from the `subscribe()` to cancel in the `stop()`
- recommended way in Reactor
* Use `onErrorContinue()` in the `FluxMessageChannel` instead of
`try..catch` in the `doOnNext()` - for possible `onErrorStop()`
in the provided upstream `Publisher`
* Handle `RMH` in the `ServiceActivatorFactoryBean` as a direct handler
as well with wrapping into `ReactiveMessageHandlerAdapter` for return.
The `ConsumerEndpointFactoryBean` extracts an `RMH` from the adapter
for the `ReactiveStreamsConsumer` anyway
* Add XML parsing test for `ReactiveMongoDbStoringMessageHandler`
* Add `log4j-slf4j-impl` for all the test runtime since `slf4j-api`
comes as a transitive dependency from many places

### CODE REVIEW

**Comment by artembilan** [comment]
TODO:

* Java DSL
* Annotations
* Docs

Although I can do them all in the separate PR(s).

@dturanski , this is what I meant about processing `ReactiveMessageHandler` in the flow.

Thanks

**Comment by artembilan** [comment]
This one depends on #3139 . So, after merging that one I'll update here.

But I definitely would like to have this merged for upcoming release.

Thanks

**Comment by garyrussell** [comment]
I will look at it today, but you might want to request a review from the Reactor team since my Reactor skills are lacking.

Minor checkstyle issue reported by Travis.

**Comment by artembilan** [comment]
> Minor checkstyle issue reported by Travis.

Will fix after merging MongoDB revert.

**Comment by artembilan** [comment]
@bsideup ,

any chances that you can sanctify this PR?

Thanks

File: spring-integration-core/src/main/java/org/springframework/integration/config/ConsumerEndpointFactoryBean.java
Comment (garyrussell) [inline spring-integration-core/src/main/java/org/springframework/integration/config/ConsumerEndpointFactoryBean.java]: Why ISE instead of IAE? This is a setter. `Assert.isTrue()` ?

File: spring-integration-core/src/main/java/org/springframework/integration/config/ConsumerEndpointFactoryBean.java
Comment (garyrussell) [inline spring-integration-core/src/main/java/org/springframework/integration/config/ConsumerEndpointFactoryBean.java]: `s/for/to a/`

**Review by garyrussell — CHANGES_REQUESTED** [review]
Minor issues; otherwise LGTM

**Comment by artembilan** [comment]
Please, find the changes for Java DSL in the latest commit.
Since no feedback from Reactor team, I guess we are good to go and can revise in the future if that.

Thanks

**Pull request merged**
